### ISSUE TITLE
Need way to add testing code for modules/apps

### ISSUE THREAD

**Issue by bwspenc** [idaholab/moose#7347]
### Description of the enhancement or error report

We need a way to add code for testing purposes for the modules and apps that is similar to what we are currently doing in framework/test.
### Rationale for the enhancement or information for reproducing the error

Sometimes testing off-normal cases requires writing objects that are for testing purposes only. There's currently nowhere to put code like that in the modules. If we add those objects to a module's directory, it will get compiled in, and someone may mistakenly use it for something other than its intended purpose.

Requirements (up for discussion, of course):
1. A standard place to put code like this that is specific to each module/app.
2. The standard module/app executable would not include this code.
3. A special testing app for each module/app that includes the testing code for that app/module, as well as the testing code for all the modules that it depends on, and the code in framework/test.
4. A place for tests in each app/module that use the corresponding testing app.
### Identified impact

Develop standardized directory scheme and makefile modificaitons.

**Comment by YaqiWang** [comment]
I like this idea. For a module, we may just need to maintain one testing executable.

**Comment by friedmud** [comment]
I like the idea. I wonder how standard it should be? I kind of want this
capability in my own Stork-based app as well. I guess I could always just
create another fork of Stork underneath my current application and link it
to my application's library.... hmmmm.

**Comment by bwspenc** [comment]
I discussed this with @permcody and @jwpeterson a little bit, and we liked the idea of having a `test` directory in every project. It would have subdirectories for `src`, `include`, and `tests`. We'd build a separate executable in there that links in all of the module code as well as the code in that `test` directory.

**Comment by friedmud** [comment]
It makes sense to me.

Are we still going to have tests in the top level as well? Or should they
all move into the testing application?

If not, then should run_tests at the application level run both sets with
each executable?

My one fear here is that we're adding more friction to creating tests.
People already don't really want to do it... and if they have to keep up
with another executable it's just one more reason to ignore testing.

Note: I'm talking about Stork-based apps here. In the framework we can
mandate whatever we want! ;-)

**Comment by jessecarterMOOSE** [comment]
Just my $0.02 worth, but as far as Stork-based apps, I find comfort in the fact that, currently, I can build my (single) executable and immediately ./run_tests and make sure I didn't hose anything up (obviously I don't have much faith in my programming ability). I have very few objects that are just for testing purposes so I don't mind them being mixed in the with "real" objects.

**Comment by brianmoose** [comment]
So the directory structure would look like:
```
- modules/phase_field
    - Makefile
    - run_tests
    - doc/
    - include/
    - lib/
    - src/
    - tests/
    - test/
        - src/
            - main.C
            - base/
                PhaseFieldTestApp.C
        - include/
            - base/
                PhaseFieldTestApp.h
        - tests/
        - Makefile
        - run_tests
```
The `Makefile` in `modules/phase_field/test/` would set up compiling with the top level framework test directory as well as any test directories (if they exist) of modules it depends on.

It makes sense to me that there would be tests for both the "test" and "non test" executables so that there would be two `tests` directories.  However, with this directory structure `modules/phase_field/run_tests` would pick up any tests in `modules/phase_field/test/tests` unless the test files were named differently. Implementing #8693 would probably help here. Or add a TestHarness option to ignore certain directories but then you would have to explicitly run both `run_tests`. 

I don't think this should be the default, probably just a script to set this up if you need it.

**Comment by permcody** [comment]
This is close but I don't think it's quite right. Yes we want a separate `test` folder underneath each module to house the test objects but I don't think we need to worry about separating out the `tests` themselves for "test" and "non test" objects. For instance, we don't have a way to test only the objects in the framework, they have to be tested in conjunction with the test application. That also means that we don't need two separate executables either. 

My thinking is that if you go into a module (let's say phase_field as an example) and build there, you will get an executable (phase_field-opt) but it'll have the test library objects linked in. This would be accomplished by building a "phase_field-opt.dylib" (as we do now) with the production objects and a "phase_field_test-opt.dylib" with just the test objects in it.

This would mean we wouldn't have to change application make files as they already are setup to just link the production libraries together. I think this just about covers all the cases.

**Comment by brianmoose** [comment]
So the point of splitting  up the libraries into "test" and "non test" is just so that apps can link against the production library and not have to use the test objects? But people using `phase_field-opt` would still see all the test objects and hopefully wouldn't abuse their use?

It seems to me there is some benefit to having two executables. The test version can have all sorts of test objects that test various aspects of the code. The production executable would just have the objects it needs and the tests would probably be more general.
That was my understanding of the original issue at least.

**Comment by permcody** [comment]
Oh right! We definitely want the subset of objects listed in the dump. We should do this for MOOSE too. We'll chat more about how we might do this next week.

**Comment by brianmoose** [comment]
After a bit of offline discussion, here are some ideas (all assuming a single executable):

1. Start off with being able to specify test only objects. We can either use a special register function, ie `registerTestKernel` or have test objects inherit from some base `TestObject`. In either case, a special run time flag would be required to be able to use test objects, something like `./phase_field-opt --test-objects`. The TestHarness would presumably add this option automatically (also possibly have the option to turn off/on on a per test basis). This would make it so that `--dump` wouldn't automatically show test only objects (without an additional command line option) and it would catch if a test object was inadvertently added to normal input files.

2. A separate issue would be to keep the test objects in a different library that is linked into the executable. This would keep test objects out of downstream usage. This would use a separate `test` directory where an automatic test library would be created. For example, if `phase_field` had a `test` directory, it would create `libphase_field-opt` and `test/libphase_field-test-opt`. Both of these would be linked into `phase_field-opt`. When building BISON, it would only link against `libphase_field-opt` and we wouldn't have to compile all the objects that would go into `libphase_field-test-opt`. 

3. A continuation of (2), it was not deemed desirable to share test objects, so there wouldn't be any automatic support for linking in other modules/apps test objects. For example, if BISON wanted to resuse a test object in phase_field, it would copy it over rather than just link in `libphase_field-test-opt`.

@permcody Anything else to add?

**Comment by permcody** [comment]
No - This captures our discussion. I wouldn't mind getting buy-in from @friedmud and @andrsd on this since neither were involved in this discussion.

**Comment by andrsd** [comment]
What's wrong with testing modules like we test MOOSE?

**Comment by permcody** [comment]
Well like with everything else, we have to weigh the costs of changing what people are use to. The framework doesn't have an executable but each module currently does. If we add a directory underneath each module would we build a different executable inside of that test directory? Would that mean we dump the existing executable? If not, then probably leave it right in the root of each module. Then the next question is how to differentiate between test objects and production objects. This is a capability we don't have in MOOSE. There's no way to dump just framework level objects, when looking at just MOOSE. If you use moose-test, you get test objects, those magically disappear when dumping from a module that doesn't link in the test app. 

There are just a lot of trade-offs to polishing this capability, so if we are going to add a test directory we want to make sure we have the right set of trade-offs. For me, I really don't want to compile all the upstream test objects (#2 above) so I think that's important to separate so that we can avoid traversing those directories and building objects in the first place.

**Comment by andrsd** [comment]
> If we add a directory underneath each module would we build a different executable inside of that test directory?

Yes.

> Would that mean we dump the existing executable?

No.

> Then the next question is how to differentiate between test objects and production objects. This is a capability we don't have in MOOSE.

Test objects are in `test`, everything else is production.

> There's no way to dump just framework level objects, when looking at just MOOSE.

The framework is not meant to be a standalone executable. Modules are apps that link the framework, they build an executable, they can produce dump with production level objects + framework objects (moose test excluded). Why do we need to dump objects just from MOOSE?

Also think about unit tests in apps, they live in `unit/`, they have their own testing classes, they build an executable and they are not part of the production executable. Currently unit tests for modules are in moose/unit 🤢

**Comment by permcody** [comment]
> Test objects are in test, everything else is production.

This is trickier to accomplish than it sounds. Inspecting paths during compilation is difficult to do in a robust way. We've learned just how bad recently when using the `__FILE__` CPP directive. Alternatively, you could register in two spots, register with a different macro, or require inheriting from some other interface to create test objects. All have pros/cons. This is the essence of what we are trying to decide here.

**Comment by friedmud** [comment]
I can definitely see all sides of this.

Personally: I think having a separate `test` directory/executable is what I'm voting for.  It's a departure from what the apps currently do - but it's cleaner and easier to manage in the long run.  It will also promote the creation of test objects which will hopefully help apps reach higher levels of test coverage.

The idea of having desperately registered test objects is decent... but just had too many downsides.  If you can't optionally compile them (like 2 says above) then we're stuck waiting on test objects to compile all the time which is a bummer.  In addition, if you can't optionally compile them then there is the large chance that two modules/apps will declare test objects named the same thing that do different things and have namespace collisions.  That would be a pretty huge disaster.

However, to get optional compiling working would be a pretty big pain.  The only way that I can see is that test objects would have to be wrapped in ifdefs.  People would forget this all the time... leading to all of the above bad behavior.

No: I think the only sane thing to do is have a separate directory with its own executable.  We really should have had it that way for a while at this point...

**Comment by permcody** [comment]
> No: I think the only sane thing to do is have a separate directory with its own executable.  We really should have had it that way for a while at this point...

So we discussed this option a lot and I'm worried it'll create confusion. There's not a hard and fast rule on when to use which executable that's easy to explain to people. You might tell people to prefer the executable sitting in the main directory but then they go build a test and figure out that they have to run the test executable. It also just feels hacky to have two sitting next to each other or even in slightly different locations. 

I suggested that we combine the behavior of the two executables into a single executable and control which objects are available for use or dumping via a command line flag (e.g. `--allow-test-objects`). This would certainly reduce possible confusion and make the appearance and delivery of our apps more pleasing. 

I was about to say that this doesn't solve the registration issue but I just say @andrsd's comment. That might just work and be completely hidden from the user... Hmm
